### ISSUE TITLE
Realm crashes in terminate_internal

### ISSUE THREAD

**Issue by Standret** [realm/realm-cocoa#6314]
## Expected Results
Working without crashes during compacting.

## Actual Results
Take a look at the stack trace.

```
0   libsystem_kernel.dylib        	0x00000001b2074ebc __pthread_kill + 8
1   libsystem_pthread.dylib       	0x00000001b1f90790 pthread_kill$VARIANT$mp + 112 (pthread.c:1456)
2   libsystem_c.dylib             	0x00000001b1ee48a0 __abort + 112 (abort.c:147)
3   libsystem_c.dylib             	0x00000001b1ee4830 abort + 112 (abort.c:118)
4   Realm                         	0x0000000105dc7c98 please_report_this_error_to_help_at_realm_dot_io + 12
5   Realm                         	0x0000000105dc7f64 realm::util::terminate_internal(std::__1::basic_stringstream<char, std::__1::char_traits<char>, s... + 264
6   Realm                         	0x0000000105dc82d4 realm::util::terminate_with_info(char const*, char const*, long, char const*, std::initializer_li... + 388
7   Realm                         	0x0000000105c0e274 void realm::util::terminate_with_info<unsigned long long const&, std::__1::basic_string<char, std... + 84
8   Realm                         	0x0000000105c0ff3c realm::SlabAlloc::do_translate(unsigned long) const + 540
9   Realm                         	0x0000000105c14f0c realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 476
10  Realm                         	0x0000000105c14fd8 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 680
11  Realm                         	0x0000000105c14fd8 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 680
12  Realm                         	0x0000000105c14fd8 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 680
13  Realm                         	0x0000000105c14fd8 realm::Array::do_write_deep(realm::_impl::ArrayWriterBase&, bool) const + 680
14  Realm                         	0x0000000105cb4e94 realm::Group::write(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, int, realm::Gro... + 244
15  Realm                         	0x0000000105cb595c realm::Group::write(realm::util::File&, char const*, unsigned long long, bool) const + 324
16  Realm                         	0x0000000105cbd928 realm::SharedGroup::compact(bool, realm::util::Optional<char const*>) + 388
17  Realm                         	0x0000000105bd3234 realm::Realm::compact() + 116 (shared_realm.cpp:782)
18  Realm                         	0x0000000105bd25e0 realm::Realm::Realm(realm::Realm::Config, std::__1::shared_ptr<realm::_impl::RealmCoordinator>) + 360 (shared_realm.cpp:81)
19  Realm                         	0x0000000105b08c50 realm::Realm::make_shared_realm(realm::Realm::Config, std::__1::shared_ptr<realm::_impl::RealmCoo... + 80 (shared_realm.hpp:364)
20  Realm                         	0x0000000105b08a70 std::__1::shared_ptr<realm::Realm::make_shared_realm(realm::Realm::Config, std::__1::shared_ptr<r... + 132 (shared_realm.hpp:364)
21  Realm                         	0x0000000105b035e8 realm::_impl::RealmCoordinator::do_get_realm(realm::Realm::Config, std::__1::shared_ptr<realm::Re... + 228 (memory:4710)
22  Realm                         	0x0000000105b03e10 realm::_impl::RealmCoordinator::get_unbound_realm() + 104 (realm_coordinator.cpp:249)
23  Realm                         	0x0000000105ba0ea4 __62+[RLMRealm asyncOpenWithConfiguration:callbackQueue:callback:]_block_invoke + 360 (RLMRealm.mm:307)
24  libdispatch.dylib             	0x00000001b1f49610 _dispatch_call_block_and_release + 24 (init.c:1408)
25  libdispatch.dylib             	0x00000001b1f4a184 _dispatch_client_callout + 16 (object.m:495)
26  libdispatch.dylib             	0x00000001b1ef2a9c _dispatch_continuation_pop$VARIANT$mp + 412 (inline_internal.h:2487)
27  libdispatch.dylib             	0x00000001b1ef21f0 _dispatch_async_redirect_invoke + 600 (queue.c:800)
28  libdispatch.dylib             	0x00000001b1efefd0 _dispatch_root_queue_drain + 376 (inline_internal.h:2528)
29  libdispatch.dylib             	0x00000001b1eff79c _dispatch_worker_thread2 + 128 (queue.c:6569)
30  libsystem_pthread.dylib       	0x00000001b1f99f64 _pthread_wqthread + 212 (pthread.c:2336)
31  libsystem_pthread.dylib       	0x00000001b1f9cae0 start_wqthread + 8
```

## Steps to Reproduce
It is reported by automatically from users.
I can only say that this type of crashes usually occurs with low-memory mode ~100MB RAM free.
I wasn't able to reproduce it by myself.

## Version of Realm and Tooling
Realm framework version: 3.18.0

Xcode version: 11.0

iOS/OSX version: 13.1.2

Dependency manager + version: Cocoapods 1.8.4

**Comment by ericjordanmossman** [comment]
The stack trace shows that this is being thrown during a write transaction. Do you have any other information on the **Steps To Reproduce**? Have you been able to reproduce this crash on your own machine? If so, can you please provide that.

Could you provide more information on the **Goal**, also? Under what specific circumstances would you like the transaction to be successful in?

**Comment by Standret** [comment]
Unfortunately, I wasn't able to reproduce it. but I can provide you with additional information, hope it will be useful.

On the application startup, we have some special code to do realm compacting and if needed record data in the realm about changes (after compacting). It starts directly on `launch` and displays progress and only after successful compacting we start application normally and begin loading data from the realm.
After App start, we start syncing local data with the server. but this part is OK and don't throw an exception

**Comment by ericjordanmossman** [comment]
@Standret Thanks for the additional information, but unfortunately we won't be able track down a problem without a reproduction for this issue.

One thing to note is that if this only happens when the device is low on memory, it may be explained depending on the realm size,

> The [compaction operation](https://realm.io/docs/swift/latest/#compacting-realms) works by reading the entire contents of the Realm file, rewriting it to a new file at a different location, then replacing the original file. Depending on the amount of data in a file, this may be an expensive operation.

**Comment by leemaguire** [comment]
Closing due to inactivity. If you have any other questions please open a new issue. Thanks.
